source/extensions/filters/network/dubbo_proxy/filter.cc

```diff
@@ -0,0 +1,228 @@
+      serialization_type_(serialization_type) {}
+
+Filter::~Filter() {}
+
+Network::FilterStatus Filter::onData(Buffer::Instance& data, bool) {
+  if (!sniffing_) {
+    if (request_buffer_.length() > 0) {
+      // Stopped sniffing during response (in onWrite). Make sure leftover request_buffer_ contents
+      // are at the start of data or the upstream will see a corrupted request.
+      request_buffer_.move(data);
+      data.move(request_buffer_);
+      ASSERT(request_buffer_.length() == 0);
+    }
+
+    return Network::FilterStatus::Continue;
+  }
+
+  ENVOY_LOG(trace, "dubbo: read {} bytes", data.length());
+  request_buffer_.move(data);
+
+  try {
+    if (!request_decoder_) {
+      request_decoder_ = createDecoder(*this, *this);
+    }
+
+    BufferWrapper wrapped(request_buffer_);
+    request_decoder_->onData(wrapped);
+
+    // Move consumed portion of request back to data for the upstream to consume.
+    uint64_t pos = wrapped.position();
+    if (pos > 0) {
+      data.move(request_buffer_, pos);
+    }
+  } catch (const EnvoyException& ex) {
+    ENVOY_LOG(error, "dubbo: error {}", ex.what());
+    data.move(request_buffer_);
+    stats_.request_decoding_error_.inc();
+    sniffing_ = false;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus Filter::onWrite(Buffer::Instance& data, bool) {
+  if (!sniffing_) {
+    if (response_buffer_.length() > 0) {
+      // Stopped sniffing during request (in onData). Make sure response_buffer_ contents are at the
+      // start of data or the downstream will see a corrupted response.
+      response_buffer_.move(data);
+      data.move(response_buffer_);
+      ASSERT(response_buffer_.length() == 0);
+    }
+
+    return Network::FilterStatus::Continue;
+  }
+
+  ENVOY_LOG(trace, "dubbo: wrote {} bytes", data.length());
+  response_buffer_.move(data);
+
+  try {
+    if (!response_decoder_) {
+      response_decoder_ = createDecoder(*this, *this);
+    }
+
+    BufferWrapper wrapped(response_buffer_);
+    response_decoder_->onData(wrapped);
+
+    // Move consumed portion of response back to data for the downstream to consume.
+    uint64_t pos = wrapped.position();
+    if (pos > 0) {
+      data.move(response_buffer_, pos);
+    }
+  } catch (const EnvoyException& ex) {
+    ENVOY_LOG(error, "dubbo: error {}", ex.what());
+    data.move(response_buffer_);
+    stats_.response_decoding_error_.inc();
+    sniffing_ = false;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+void Filter::onEvent(Network::ConnectionEvent event) {
+  if (active_call_map_.empty()) {
+    return;
+  }
+
+  if (event == Network::ConnectionEvent::RemoteClose) {
+    stats_.cx_destroy_local_with_active_rq_.inc();
+  }
+
+  if (event == Network::ConnectionEvent::LocalClose) {
+    stats_.cx_destroy_remote_with_active_rq_.inc();
+  }
+}
+
+void Filter::onRequestMessage(RequestMessagePtr&& message) {
+  if (!message) {
```

Comment:
shouldn't this two be `ASSERT`? i.e. it is certainly a bug in the proxy code so shouldn't be caught but crash.

Reply:
done